api/spec/v1.yaml

```diff
@@ -961,6 +961,17 @@ paths:
             A filter on the evm log signatures.
             Note: The filter will only match on parsed (verified) EVM events.
           example: '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef'
+        - in: query
+          name: evm_log_contract_addr_eth
+          schema:
+            type: string
+          description: Get EVM log events from the contract at this Ethereum address.
+        - in: query
+          name: erc721_transfer_token_id
+          schema:
+            type: string
+          example: "999"
+          description: Get ERC-721 Transfer events about this token ID.
```

Comment:
It feels wrong to couple the type of the event (ERC-721 transfer) and the token ID so tightly in a single filter. Especially since we already have the `evm_log_signature` filter. 
I suggest we rename this simply `nft_token_id`, and adjust the description accordingly: "A filter on the token (instance) ID of an NFT. All returned events will involve this token." This leaves the API more amenable to future changes.

I'm guessing you bundled these two together due to performance concerns? I share those :). We can add to the description: "If this parameter is provided, `evm_log_contract_addr_eth` has to be provided too, and `evm_log_signature` has to equal 0x... (the ERC-721 Transfer event)". We can then enforce this at the param validation layer, before we call `RuntimeEvents()`. This puts a tiny bit more onus on the user, but makes the query more explicit, and keeps it possible for us to relax the constraint in the future, should we want to (ERC-1155 support, Cipher support, ...), without having to deprecate this query param and invent new ones.

Reply:
correctness:

- the implementation behind this literally only looks at ERC-721 transfers. and although ERC-721 is the only supported standard which would waive us having to specify ERC-721 here, it doesn't look at other event types from ERC-721, including Approval, and debatably, ApprovalForAll.

maintainability:

- if we add an NFT standard other than ERC-721 without broadening this, this name will become wrong

Reply:
I don't think I get your points above. They sound like summaries of my comment to me :)
- correctness: Agreed. Hence my suggestion to "add to the description" (of the param).
- maintainability: Also agreed, hence my suggestion to rename the param.

Regardless, thank you for the changes. I still think we should enforce at the API level that if `nft_id` is given, then `contract_address` has to be given, and (as you point out) `evm_log_signature` has to be `0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef`, for two reasons:
- Explicitness/Correctness. If users provide a different `evm_log_signature` (e.g. for Approval), we won't correctly return those events with the current code; the SQL query hardcodes the Transfer signature.
- **Preventing expensive queries. If people request all events with `ntf_id==0` (with no `contract_address` specified), that will result in a very expensive linear scan of the events.** We can fight that by adding an index (and then continuing to ensure that this wacko query is performant, yuck), or by preventing this type of queries.

Reply:
OK! Yeah if it's not technically more difficult to support arbitrary event signatures, great, let's do that. :+1: 

I was suggesting we could demand the specific ERC721 Transfer signature _for now_ because that's all that our Explorer needs (it has a dedicated tab or showing past token transfers) and because limiting the scope of what we're able to retrieve efficiently can make the task easier. In particular, I thought we might need to use the commented-out `CREATE INDEX ...WHERE evm_log_signature=0xddf...`.

Reply:
ok I'll test that on staging and uncomment it if it works

Reply:
wait did I reply to the wrong thread aaaaa

Reply:
deleted replies to wrong thread

the index is un-commented-out. we're currently not forcing the user to set $6 = (that event signature hash). we can later add `Allow` events similarly if desired

Reply:
reinstating specifying the _current_ limitation to ERC-721 Transfer events

Reply:
adding a note that users should specify evm_log_signature if they want to get only Transfer, in case we add support for Approval later